[21.05] FDE: admin fingerprint

osnyx opened the pull request:
@flyingcircusio/release-managers

Compute and display a hash fingerprint of the LUKS admin passphrase entered and store it alongside the key file on the keystore USB stick.
This helps discovering accidentally wrong keys entered into a system and debug unknown/ corrupted passphrase states.

Missing fingerprints or mismatches are displayed and resolved interactively.

Includes subcommand `fc-luks keystore fingerprint` for calculating the fingerprint of a passphrase and then storing it in a vault of your choice.

PL-131325

## Release process

Impact: internal

Changelog:
- fc-ceph/ fc-luks: entering an admin key now displays a fingerprint of the correct key on mismatches
- `fc-luks keystore fingerprint`

### PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [x] ticket state set to *Pull request ready*
- [x] if ticket is more urgent than within the next few days, directly contact a member of the Platform team

## Design notes

- [x] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
- [x] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - [x] improve handling and discoverability of passphrase mismatches for operators
  - [x] extend automated tests
  - [x] fine-grained unit tests for all common input flows
  - [x] must not introduce known regressions
  - [x] must work on existing encrypted hosts and update them with a fingerprint
- [x] Security requirements tested? (EVIDENCE)
  - [x] automated tests pass
  - [x] extended unit tests
  - [x] manually tested in dev cluster
  - [x] manually initialised the fingerprint on host while maintaining the same passphrase (migration scenario)

osnyx committed:
fc-ceph: admin passphrase fingerprint implementation

Core omplementation of generating a cryptographic fingerprint of the
admin key, persisting it in the keystore, comparing it at unlock time
and displaying/ handling mismatches.

Includes unit tests and adjusted NixOS tests:
Play through all cases of entering a correct or incorrect admin
passphrase, dealing with fingerprint mismatches, updates, using the
entered passphrase anyways, and retrying on confirmation mismatches.

PL-131325

Co-author: Christian Theune <ct@flyingcircus.io>

osnyx committed:
fc-ceph Mdraid: use defined disk constants also for list slicing

osnyx committed:
py_pytest_patterns: Move to overlay for further use

We now use the pytest-patterns plugin in multiple packages – but also with multiple python versions from different nixpkgs revisions.
This is why I've decided to not properly integrate it into the python3Packages set yet.

osnyx committed:
fc-luks rekey: fix external header heuristic being to broad

Accidentally, an external header discovered using the heuristics was used for all volumes,
even if they did not require an external header.

ctheune commented on pkgs/fc/ceph/src/fc/ceph/luks/__init__.py:
```diff
@@ -1,24 +1,126 @@
+
+
+class FingerprintedPassphrase:
+    class Fingerprint(bytes):
```
move to top level scope

ctheune commented on pkgs/fc/ceph/src/fc/ceph/luks/__init__.py:
```diff
@@ -89,7 +250,8 @@ def luksFormat(cls, *args: str, **kwargs):
             *cls._tunables_cipher,
             "luksFormat",
-            *args, **kwargs,
+            *args,
```
your formatter ignored the `fmt: off` directives

osnyx replied:
No, *someone's* formatter did ;)

ctheune commented on pkgs/fc/ceph/src/fc/ceph/luks/manage.py:
```diff
@@ -97,33 +97,38 @@ def rekey(
             run.json.lsblk("-s", "-o", "NAME,PATH,TYPE,MOUNTPOINT")
         )
         for candidate in candidates:
+            this_header = header
```
we shouldn't intermingle giving a specific luks header to a multi-volume operation

osnyx committed:
fc-ceph: remove Experimental label from encryption support

osnyx committed:
fc-luks keystore fingerprint command

includes unit tests

ctheune commented on pkgs/fc/ceph/src/fc/ceph/luks/__init__.py:
```diff
@@ -1,24 +1,33 @@
 
 class LUKSKeyStore(object):
-    __admin_key = None
+    __admin_key: Optional[bytes] = None
```
this is unused, or rather should be __admin_key_for_input

ctheune committed:
upgrade black to ensure black/ruff compatibility

ctheune committed:
review: small simplifications, some formatter stuff, some typos

ctheune reviewed (changes requested):
overall: very good, the one thing about handling external headers vs. mass operations needs an alternative solution

osnyx committed:
fc-luks: require specific device when rekeying with external header

External LUKS headers belong to a single LUKS volume only.
To avoid accidental errors, require that only 1 device is matched
when rekeying with an external header file.

This does not affect the auto-discovered external headers.
